src/transformers/models/emformer/modeling_emformer.py

```diff
@@ -0,0 +1,1158 @@
+
+        # Compute attention probabilities.
+        attention_probs = self._gen_attention_probs(attention_weights, attention_mask, padding_mask)
+        if output_attentions:
+            attention_outputs = attention_probs.reshape(batch_size, self.num_heads, time_length, time_length)
+            utterance_start = right_context.size(0)
+            utterance_end = right_context.size(0) + utterance.size(0)
+            attention_outputs = attention_outputs[:, :, utterance_start:utterance_end, utterance_start:utterance_end]
+        else:
+            attention_outputs = None
+
+        # Compute attention.
+        attention = torch.bmm(attention_probs, reshaped_value)
+        # attention.shape == (batch_size * self.num_heads, time_length, self.input_dim // self.num_heads)
+        attention = attention.transpose(0, 1).contiguous().view(time_length, batch_size, self.input_dim)
+
+        # Apply output projection.
+        output_right_context_mems = self.out_proj(attention)
+
+        summary_length = summary.size(0)
+        output_right_context = output_right_context_mems[: time_length - summary_length]
+        output_mems = output_right_context_mems[time_length - summary_length :]
+        if self.tanh_on_mem:
+            output_mems = torch.tanh(output_mems)
+        else:
+            output_mems = torch.clamp(output_mems, min=-10, max=10)
+
+        return output_right_context, output_mems, key, value, attention_outputs
+
+    def forward(
+        self,
+        utterance: torch.Tensor,
+        lengths: torch.Tensor,
+        right_context: torch.Tensor,
+        summary: torch.Tensor,
+        mems: torch.Tensor,
+        attention_mask: torch.Tensor,
+        output_attentions: bool = False,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        r"""Forward pass for training.
+
+        B: batch size; D: feature dimension of each frame; T: number of utterance frames; R: number of right context
+        frames; S: number of summary elements; M: number of memory elements.
+
+        Args:
+            utterance (torch.Tensor): utterance frames, with shape *(T, B, D)*.
+            lengths (torch.Tensor): with shape *(B,)* and i-th element representing
+                number of valid frames for i-th batch element in `utterance`.
+            right_context (torch.Tensor): right context frames, with shape *(R, B, D)*.
+            summary (torch.Tensor): summary elements, with shape *(S, B, D)*.
+            mems (torch.Tensor): memory elements, with shape *(M, B, D)*.
+            attention_mask (torch.Tensor): attention mask for underlying attention module.
+
+        Returns:
+            (Tensor, Tensor):
+                Tensor
+                    output frames corresponding to utterance and right_context, with shape *(T + R, B, D)*.
+                Tensor
+                    updated memory elements, with shape *(M, B, D)*.
+        """
+        output, output_mems, _, _, attention_probs = self._forward_impl(
```

Comment:
Why do you return 3, 4 if it's never used? Let's just not return it in this case no?